src/OpenTelemetry/Trace/ExceptionProcessor.cs

```diff
@@ -19,7 +21,14 @@ internal sealed class ExceptionProcessor : BaseProcessor<Activity>
     public ExceptionProcessor()
     {
 #if NET6_0_OR_GREATER || NETFRAMEWORK
-        this.fnGetExceptionPointers = Marshal.GetExceptionPointers;
+        if (RuntimeFeature.IsDynamicCodeSupported)
```

Comment:
```suggestion
        bool getExceptionPointersSupported = true;
        try { Marshal.GetExceptionPointers(); } catch { getExceptionPointersSupported = false; }
        if (!getExceptionPointersSupported)
            throw new throw new NotSupportedException($"'{typeof(Marshal).FullName}.GetExceptionPointers' is not supported.");
```
Instead of calling `RuntimeFeature.IsDynamicCodeSupported`, can this just call the API to see whether it is supported?

Note that this API is also not supported in Mono (irrespective of whether IsDynamicCodeSupported is true or false), so it would cover the Mono.

Reply:
> what would it take to actually support the API in native AOT?
> 
> Yes, it should be possible to simulated the API in native AOT. The API would need to allocate a piece of memory with the same lifetime as the current exception in flight (if it was not previously allocated), fill it with gunk, and return the pointer to it. The "new" exception handling that is being enabled for CoreCLR and that is not internally based on SEH proves that it is possible.
> 
> (Also, as I have said earlier - introducing a new API that is explicitly designed for this scenario would be the proper solution.)

Thank you! I agree above is the proper way to handle the situation. shall I log this feature request in the runtime repo?

Reply:
> Shall I log this feature request in the runtime repo?

Yes, please.

Reply:
I've filed a request to implement the API in runtime: https://github.com/dotnet/runtime/issues/98878.
Chatted with @eerhardt, @jkotas and @reyang offline.

While the above API is being implemented, we may do the followings to avoid crashing the consumer's app:
1. Annotate `SetErrorStatusOnException()` with [RequiresDynamicCode] to serve as a compile time warning for the consumer. One caveat of this approach is IsDynamicCode does not really explain the nature of this issue and may further confuse the consumer.
2. Use try/catch to swallow the exception if users hit `System.PlatformNotSupportedException` upon calling `Marshal.GetExceptionPointer()`. Store the state in `ExceptionProcessor()` and dump the Message "Exception processor not supported" inside `OnEnd(Activity activity)`.

Not sure which one is a better stopgap of this issue?
cc @MichalStrehovsky, @agocke

Reply:
Discussed in today's SIG meeting.
I'll go with adding the compile time warning as was mentioned in 1. above.
Plus, guarding the extension method https://github.com/open-telemetry/opentelemetry-dotnet/blob/e7cbbbbf0d3db71662eb01cee5285ed613339ea2/src/OpenTelemetry/Trace/Builder/TracerProviderBuilderSdk.cs#L144
with a try/catch on Marshal.GetExceptionPointers to avoid adding useless Processor at runtime (if user ever decided to suppress the compile time AOT warning.)

Reply:
I'm OK with either re-throwing or just silently failing & not adding the processor. Probably re-throwing is better to alert the users 🤔

Reply:
+1 on re-throwing the exception at start-up.